notebook/notebookapp.py

```diff
@@ -1308,14 +1299,15 @@ def parse_command_line(self, argv=None):
             self.update_config(c)
 
     def init_configurables(self):
-        self.kernel_spec_manager = self.kernel_spec_manager_class(
-            parent=self,
-        )
+        if self.kernel_providers:
+            self.kernel_finder = KernelFinder(self.kernel_providers)
+        else:
+            self.kernel_finder = KernelFinder.from_entrypoints()
```

Comment:
This seems a little awkward to me.  How would an application be able to have providers from both the property and the entrypoints?  If there was a way to get a list of entrypoint-based providers, then you could require the property be the composite.  

If there was a separate way to get the providers from entrypoints, then you could just call the constructor and, if the argument is None, have it call `from_entrypoints()`, else the argument is the complete set.

Reply:
Sounds good.  I would recommend these approaches get pushed into the constructor and let it determine how to compose/filter the providers.